java/statestore/src/main/java/sleeper/statestore/dynamodb/DynamoDBFileReferenceStore.java

```diff
@@ -304,22 +392,37 @@ public void deleteGarbageCollectedFileReferenceCounts(List<String> filenames) th
         double totalCapacityConsumed = 0;
         double batchCapacityConsumed = 0;
         for (String filename : filenames) {
-            DeleteItemRequest delete = new DeleteItemRequest().withTableName(fileReferenceCountTableName)
+            TransactWriteItem writeItem = new TransactWriteItem().withDelete(new Delete()
+                    .withTableName(fileReferenceCountTableName)
                     .withKey(fileReferenceFormat.createReferenceCountKey(filename))
                     .withConditionExpression("#References = :refs")
                     .withExpressionAttributeNames(Map.of("#References", REFERENCES))
-                    .withExpressionAttributeValues(Map.of(":refs", createNumberAttribute(0)));
+                    .withExpressionAttributeValues(Map.of(":refs", createNumberAttribute(0)))
+                    .withReturnValuesOnConditionCheckFailure(ReturnValuesOnConditionCheckFailure.ALL_OLD));
             try {
-                DeleteItemResult result = dynamoDB.deleteItem(delete);
+                TransactWriteItemsResult result = dynamoDB.transactWriteItems(new TransactWriteItemsRequest().withTransactItems(writeItem));
                 if (result.getConsumedCapacity() != null) {
-                    batchCapacityConsumed += result.getConsumedCapacity().getCapacityUnits();
+                    batchCapacityConsumed += result.getConsumedCapacity().get(0).getCapacityUnits();
                 }
                 if (i % 100 == 0) {
                     LOGGER.debug("Deleted 100 unreferenced files, capacity consumed = {}", batchCapacityConsumed);
                     totalCapacityConsumed += batchCapacityConsumed;
                     batchCapacityConsumed = 0;
                 }
-            } catch (AmazonDynamoDBException e) {
+            } catch (TransactionCanceledException e) {
+                List<CancellationReason> reasons = e.getCancellationReasons();
+                for (int j = 0; j < reasons.size(); j++) {
+                    CancellationReason reason = reasons.get(j);
+                    if (isConditionCheckFailure(reason)) {
+                        if (reason.getItem() == null) {
+                            throw new FileNotFoundException(filenames.get(j));
+                        } else {
+                            throw new FileHasReferencesException(
+                                    getStringAttribute(reason.getItem(), FILENAME),
+                                    getIntAttribute(reason.getItem(), REFERENCES, 0));
+                        }
+                    }
+                }
```

Comment:
This can be its own method to process the cancellation reasons, or split further. We can avoid discarding the original TransactionCanceledException.